core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -390,6 +391,18 @@ public void shutdown(Status reason) {
     }
   }
 
+  @Override
+  public String toString() {
+    Object addressGroupCopy;
+    synchronized (lock) {
+      addressGroupCopy = addressGroup.toString();
+    }
+    return MoreObjects.toStringHelper(this)
+          .add("logId", logId.getId())
+          .add("addressGroup", addressGroupCopy.toString())
```

Comment:
nit: no need to call `toString()` explicitly either, toStringHelper will do that for you.